vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/PollItemFactory.kt

```diff
@@ -0,0 +1,135 @@
+import im.vector.app.features.home.room.detail.timeline.item.PollOptionViewState
+import im.vector.app.features.home.room.detail.timeline.item.PollResponseData
+import im.vector.app.features.poll.PollState
+import im.vector.lib.core.utils.epoxy.charsequence.toEpoxyCharSequence
+import org.matrix.android.sdk.api.extensions.orFalse
+import org.matrix.android.sdk.api.session.room.model.message.MessagePollContent
+import org.matrix.android.sdk.api.session.room.model.message.PollAnswer
+import javax.inject.Inject
+
+class PollItemFactory @Inject constructor(
+        private val stringProvider: StringProvider,
+        private val avatarSizeProvider: AvatarSizeProvider,
+        private val colorProvider: ColorProvider,
+        private val dimensionConverter: DimensionConverter,
+) {
+
+    fun create(
+            pollContent: MessagePollContent,
+            informationData: MessageInformationData,
+            highlight: Boolean,
+            callback: TimelineEventController.Callback?,
+            attributes: AbsMessageItem.Attributes,
+    ): VectorEpoxyModel<*>? {
+        val pollResponseSummary = informationData.pollResponseAggregatedSummary
+        val pollState = createPollState(informationData, pollResponseSummary, pollContent)
+        val pollCreationInfo = pollContent.getBestPollCreationInfo()
+        val questionText = pollCreationInfo?.question?.getBestQuestion().orEmpty()
+        val question = createPollQuestion(informationData, questionText, callback)
+        val optionViewStates = pollCreationInfo?.answers?.mapToOptions(pollState, informationData)
+        val totalVotesText = createTotalVotesText(pollState, pollResponseSummary)
+
+        return PollItem_()
+                .attributes(attributes)
+                .eventId(informationData.eventId)
+                .pollQuestion(question)
+                .canVote(pollState.isVotable())
+                .totalVotesText(totalVotesText)
+                .optionViewStates(optionViewStates)
+                .edited(informationData.hasBeenEdited)
+                .highlighted(highlight)
+                .leftGuideline(avatarSizeProvider.leftGuideline)
+                .callback(callback)
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun createPollState(
+            informationData: MessageInformationData,
+            pollResponseSummary: PollResponseData?,
+            pollContent: MessagePollContent,
+    ): PollState = when {
+        !informationData.sendState.isSent() -> PollState.Sending
+        pollResponseSummary?.isClosed.orFalse() -> PollState.Ended
+        pollContent.getBestPollCreationInfo()?.isUndisclosed().orFalse() -> PollState.Undisclosed
+        pollResponseSummary?.myVote?.isNotEmpty().orFalse() -> PollState.Voted(pollResponseSummary?.totalVotes ?: 0)
+        else -> PollState.Ready
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun List<PollAnswer>.mapToOptions(
```

Comment:
rather than exposing the internals for testing I would take a step back and consider what we can and can't easily unit test, in this case it's the `AndroidViews/Epoxy Items`

which means instead of returning an epoxy `PollItem`, we could create a dedicated proxy `ViewState` data class for the `PolItem` 

``` kotlin
// MessageItemFactory

createPollItem() {
  val viewState = pollItemViewStateFactory.create()
  return PollItem_()
    .callback()
    .highlight()
    .state(viewState)
}
```

If `Views` have no/minimal business logic, then it means our `ViewStates` could be considered UI proxies (which means less need for slow/flaky instrumentation tests)

whichhhh would turn out class api into...

```kotlin
val viewState = PollItemViewStateFactory().create(
  pollContent: MessagePollContent,
  informationData: MessageInformationData,
)
```

as the factory returns immutable data we can make simple asserts against the result 

```kotlin
@Test
fun `given a certain input, when creating state, then has specific result`() {
  val specific: MessagePollContent

  val result = factory.create(specificInput) 

  result shouldBeEqualTo expectedOutput
} 
```

Reply:
Very nice idea, done!